src/Cryptol/Eval/FFI.hs

```diff
@@ -153,13 +149,23 @@ foreignPrim name FFIFunType {..} impl tenv = buildFun ffiArgTypes []
   marshalArg (FFIBasic t) val f = getMarshalBasicArg t \m -> do
     arg <- m val
     f [SomeFFIArg arg]
-  marshalArg (FFIArray (evalFinType -> n) t) val f =
-    getMarshalBasicArg t \m -> do
-      args <- traverse (>>= m) $ enumerateSeqMap n $ fromVSeq val
-      -- Since we need to do an Eval action in an IO callback, we need to
-      -- manually unwrap and wrap the Eval datatype
+  marshalArg (FFIArray (map evalFinType -> sizes) t) val f =
+    getMarshalBasicArg t \m ->
+      -- Since we need to do Eval actions in an IO callback, we need to manually
+      -- unwrap and wrap the Eval datatype
       Eval \stk ->
-        withArray args \ptr ->
+        allocaArray (fromInteger $ product sizes) \ptr -> do
```

Comment:
The loop here is a bit hard to follow.  I wonder if something like this (untested), might be better:
```
marshallArray [] _ _ = error "bad"

marshallArray [_] ptr v  =
  do runEval stk (m v) >>= pokeElemOff ptr
     pure (plusPtr ptr 1)       

marshallArray (_ : next) ptr v =
  do vs' <- traverse (runEval stk) $ enumerateSeqMap n $ fromVSeq v
     foldM (marshallArray next) ptr vs'
```

Reply:
Yeah, I initially tried to write it in a tail-recursive way, but I realized it probably doesn't make much difference in terms of performance.